presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterManager.java

```diff
@@ -13,81 +13,205 @@
         RouterSpec routerSpec = parseRouterConfig(config)
                 .orElseThrow(() -> new PrestoException(CONFIGURATION_INVALID, "Failed to load router config"));
+        Map<String, GroupSpec> groups = ImmutableMap.copyOf(routerSpec.getGroups().stream().collect(toMap(GroupSpec::getName, group -> group)));
+        List<SelectorRuleSpec> groupSelectors = ImmutableList.copyOf(routerSpec.getSelectors());
+        Scheduler scheduler = new SchedulerFactory(routerSpec.getSchedulerType()).create();
+        this.currentConfig = new AtomicReference<>(new ClusterManagerConfig(groups, groupSelectors, scheduler));
+        this.remoteInfoFactory = requireNonNull(remoteInfoFactory, "remoteInfoFactory is null");
+        this.initializeServerWeights();
+        this.initializeMembersDiscoveryURI();
+        List<URI> allClusters = getAllClusters();
+        allClusters.forEach(uri -> {
+            log.info("Attaching cluster %s to the router", uri.getHost());
+            remoteClusterInfos.put(uri, remoteInfoFactory.createRemoteClusterInfo(discoveryURIs.get(uri)));
+            remoteQueryInfos.put(uri, remoteInfoFactory.createRemoteQueryInfo(discoveryURIs.get(uri)));
+            log.info("Successfully attached cluster %s to the router. Queries will be routed to cluster after successful health check", uri.getHost());
+        });
+        pollingInterval = remoteStateConfig.getPollingInterval();
+        onConfigChangeDetection = () -> {
+            RouterSpec updateRouterSpec = parseRouterConfig(routerConfig)
+                    .orElseThrow(() -> new PrestoException(CONFIGURATION_INVALID, "Failed to load router config"));
+            Map<String, GroupSpec> newGroups = ImmutableMap.copyOf(updateRouterSpec.getGroups().stream().collect(toMap(GroupSpec::getName, group -> group)));
+            List<SelectorRuleSpec> newGroupSelectors = ImmutableList.copyOf(updateRouterSpec.getSelectors());
+            Scheduler newScheduler = new SchedulerFactory(updateRouterSpec.getSchedulerType()).create();
+            this.currentConfig.set(new ClusterManagerConfig(newGroups, newGroupSelectors, newScheduler));
+            this.initializeServerWeights();
+            this.initializeMembersDiscoveryURI();
+            List<URI> updatedAllClusters = getAllClusters();
 
-        this.groups = ImmutableMap.copyOf(routerSpec.getGroups().stream().collect(toMap(GroupSpec::getName, group -> group)));
-        this.groupSelectors = ImmutableList.copyOf(routerSpec.getSelectors());
-        this.schedulerType = routerSpec.getSchedulerType();
-        this.scheduler = new SchedulerFactory(routerSpec.getSchedulerType()).create();
+            updatedAllClusters.forEach(uri -> {
+                if (!remoteClusterInfos.containsKey(uri)) {
+                    log.info("Attaching cluster %s to the router", uri.getHost());
+                    remoteClusterInfos.put(uri, this.remoteInfoFactory.createRemoteClusterInfo(discoveryURIs.get(uri)));
```

Comment:
The creating and use of `discoveryURIs` is limited to just this `onConfigChangeDetection` method. There is no need to store it as a member of the ClusterManager class